src/posix/platform/trel.cpp

```diff
@@ -0,0 +1,546 @@
+    // service registration in the case of a non-mDNS local DNS-SD
+    // service.
+    //
+    // A subsequent call to this function updates the previous service.
+    // It is used to update the TXT record data and/or the port
+    // number.
+    //
+    // The `aTxtData` buffer is not persisted after the return from this
+    // function. The platform layer MUST not keep the pointer and
+    // instead copy the content if needed.
+
+    OT_UNUSED_VARIABLE(aPort);
+    OT_UNUSED_VARIABLE(aTxtData);
+    OT_UNUSED_VARIABLE(aTxtLength);
+}
+
+OT_TOOL_WEAK void trelDnssdRemoveService(void)
+{
+    // This function removes any previously registered "_trel._udp"
+    // service using `platTrelRegisterService()`. Device must stop
+    // advertising TREL service after this call.
+}
+
+OT_TOOL_WEAK void trelDnssdUpdateFdSet(fd_set *aReadFdSet, fd_set *aWriteFdSet, int *aMaxFd, struct timeval *aTimeout)
+{
+    // This function can be used to update the file descriptor sets
+    // by DNS-SD layer (if needed).
+
+    OT_UNUSED_VARIABLE(aReadFdSet);
+    OT_UNUSED_VARIABLE(aWriteFdSet);
+    OT_UNUSED_VARIABLE(aMaxFd);
+    OT_UNUSED_VARIABLE(aTimeout);
+}
+
+OT_TOOL_WEAK void trelDnssdProcess(otInstance *aInstance, const fd_set *aReadFdSet, const fd_set *aWriteFdSet)
+{
+    // This function performs processing by DNS-SD (if needed).
+
+    OT_UNUSED_VARIABLE(aInstance);
+    OT_UNUSED_VARIABLE(aReadFdSet);
+    OT_UNUSED_VARIABLE(aWriteFdSet);
+}
+
+//---------------------------------------------------------------------------------------------------------------------
+// otPlatTrel
+
+void otPlatTrelEnable(otInstance *aInstance, uint16_t *aUdpPort)
+{
+    OT_UNUSED_VARIABLE(aInstance);
+
+    assert(sInitialized);
+
+    VerifyOrExit(!sEnabled);
+
+    PrepareSocket(*aUdpPort);
+    trelDnssdStartBrowse();
+
+    sEnabled = true;
+
+exit:
+    return;
+}
+
+void otPlatTrelDisable(otInstance *aInstance)
+{
+    OT_UNUSED_VARIABLE(aInstance);
+
+    assert(sInitialized);
+    VerifyOrExit(sEnabled);
+
+    close(sSocket);
+    sSocket = -1;
+    trelDnssdStopBrowse();
+    trelDnssdRemoveService();
+
+exit:
+    return;
+}
+
+void otPlatTrelSend(otInstance *      aInstance,
+                    const uint8_t *   aUdpPayload,
+                    uint16_t          aUdpPayloadLen,
+                    const otSockAddr *aDestSockAddr)
+{
+    OT_UNUSED_VARIABLE(aInstance);
+
+    VerifyOrExit(sEnabled);
+
+    assert(aUdpPayloadLen <= TREL_MAX_PACKET_SIZE);
+
+    // We try to send the packet immediately. If it fails (e.g.,
+    // network is down) `SendPacket()` returns `OT_ERROR_ABORT`. If
+    // the send operation would block (e.g., socket is not yet ready
+    // or is out of buffer) we get `OT_ERROR_INVALID_STATE`. In that
+    // case we enqueue the packet to send it later when socket becomes
+    // ready.
+
+    if (SendPacket(aUdpPayload, aUdpPayloadLen, aDestSockAddr) == OT_ERROR_INVALID_STATE)
```

Comment:
I wonder if we should always enque if the queue is not empty, so that the packets are always sent in order.

Reply:
Good idea. I changed the code to check if there is anything in the queue we add new one in the queue as well
(i.e. we try to send immediately only if the queue is empty).